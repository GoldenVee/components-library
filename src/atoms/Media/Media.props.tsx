import { ComponentPropsWithRef } from 'react';

export interface MediaProps extends ComponentPropsWithRef<'img'> {
  title?: string;
  mediaSrc?: string;
  altText?: string;
  width?: string;
  children?: React.ReactNode;
}
