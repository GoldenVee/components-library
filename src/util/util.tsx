import React from 'react';
import { ThemeProvider } from 'styled-components';
import { themes } from '../themes';

export const ThemeWrapped = (Story: React.ReactElement): React.ReactElement => {
  return (
    <ThemeProvider theme={themes['light'] || themes['dark']}>
      {Story}
    </ThemeProvider>
  );
};
