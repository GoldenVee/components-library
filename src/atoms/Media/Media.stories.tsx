import React from 'react';
import { Meta, Story } from '@storybook/react';
import { Media } from './Media';
import { MediaProps } from './Media.props';
import styled from 'styled-components';
import geometric from '../../assets/images/geometric.png';
import maxwellTheCat from '../../assets/images/maxwell-the-cat.gif';

const meta: Meta = {
  title: 'atoms/Media',
  component: Media,
  argTypes: {
    mediaSrc: {
      description: 'MediaSrc can accept gifs, pngs, svg, and webp.',
      control: {
        type: 'text',
      },
    },
    altText: {
      description:
        'Alt text description for accessibility. If the image is decorative, leave the alt text blank.',
      control: {
        type: 'text',
      },
    },
    width: {
      description:
        'Determine the size of the image via width. Default values are auto, icon, xs, small, medium, mlg, large, xl, and full. But can accept any image size.',
      control: {
        type: 'text',
      },
    },
  },
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const ExampleContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  flex-wrap: wrap;
  height: fit-content;
  width: 100%;
  max-width: 1000px;
`;

const Template: Story<MediaProps> = (args) => <Media {...args} />;
const ExampleTemplate: Story<MediaProps> = (args) => (
  <ExampleContainer>
    <div>
      <p style={{ marginBottom: '16px' }}>Local image:</p>
      <Media mediaSrc={geometric} altText="cat sleeping on robot" {...args} />
    </div>
    <div>
      <p style={{ marginBottom: '16px' }}>Url image:</p>
      <Media
        mediaSrc="https://media.sketchfab.com/models/26f966d4a8e34828a36058ddd6b8a72a/thumbnails/c1d0e2e95cfd4fa5ab49456b47780448/18c761a640ec4e13b18adc9ff17f94fb.jpeg"
        altText="puppycat"
        width="medium"
        {...args}
      />
    </div>
    <div>
      <p style={{ marginBottom: '16px' }}>Local gif:</p>
      <Media
        mediaSrc={maxwellTheCat}
        altText="dancing cat gif"
        width="small"
        {...args}
      />
    </div>
  </ExampleContainer>
);

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
  width: 'ml',
};
export const Examples = ExampleTemplate.bind({});
Examples.args = {};
