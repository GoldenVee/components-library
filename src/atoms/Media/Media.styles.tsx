import styled, { css } from 'styled-components';
import tw, { TwStyle } from 'twin.macro';
import { MediaProps } from './Media.props';

const widthStyles = () => css`
  width: ${({ width = '1/2' }: MediaProps) => (width ? Sizes[width] : width)};
`;

export const StyledMedia = styled.img<MediaProps>(() => [
  tw`max-w-screen-2xl h-min`,
  widthStyles(),
]);

export const Sizes: TwStyle = {
  auto: 'auto',
  icon: '1.2rem',
  xs: '4.6rem',
  small: '10rem',
  ml: '11.25rem',
  medium: '20rem',
  mlg: '30rem',
  large: '45rem',
  xl: '60rem',
  full: '100%',
};
