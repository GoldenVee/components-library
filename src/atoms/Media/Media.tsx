import React from 'react';
import { MediaProps } from './Media.props';
import { StyledMedia } from './Media.styles';

export const Media = ({
  title,
  mediaSrc = 'https://via.placeholder.com/600x400/364987/ffffff?text=Placeholder',
  altText = 'placeholder image',
  width = 'ml',
}: MediaProps): JSX.Element => {
  return (
    <StyledMedia
      title={title}
      src={mediaSrc}
      alt={altText}
      width={width}
      role="img"
    />
  );
};

Media.displayName = 'Media';
