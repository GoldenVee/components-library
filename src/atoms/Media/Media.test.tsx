import React from 'react';
import { render } from '@testing-library/react';
import { Default } from './Media.stories';
import { ThemeWrapped } from '../../util/util';

describe('Media', () => {
  it('renders as expected', () => {
    const { container } = render(ThemeWrapped(<Default title="test" />));
    expect(container).toMatchSnapshot();
  });
});
