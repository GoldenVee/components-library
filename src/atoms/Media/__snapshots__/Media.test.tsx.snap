// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media renders as expected 1`] = `
<div>
  <img
    alt="placeholder image"
    class="Mediastyles__StyledMedia-sc-jkb874-0 jdBcDx"
    role="img"
    src="https://via.placeholder.com/600x400/364987/ffffff?text=Placeholder"
    title="test"
    width="ml"
  />
</div>
`;
